### PR TITLE
[Feature]heterograph.edata and heterograph.ndata support multiple ntype and etype

### PR DESCRIPTION
**Pull request by classicsong**
## Description
When heterograph has multiple node types or edge types, it is not allowed to use g.ndata['h'] or g.edata['h'] to get data with key 'h'.

So it is impossible to do following to get a edge subgraph:
```
train_mask = g.edata['train_mask']
train_g = g.edge_subgraph(train_mask)
```

We need to do the following:
```
train_edges = {}
for etype in g.canonical_etypes:
    mask = g.edges[etype].data['train_mask']
    train_edges[etype] = th.arange(g.number_of_edges(etype))[mask]
train_g = g.edge_subgraph(train_edges)
```

Another benefit can be, for model like GCMC:
```
for ntype in enc_g.ntypes:
    feat= enc_g.nodes[ntype].ndata['h']
    dec_g.nodes[ntype].ndata['h'] = feat
```
can be written as:
```
dec_g.ndata['h'] = enc_g.ndata['h']
```

This PR will support g.ndata['h'] or g.edata['h'] return a dictionary of node/edge types with their corresponding data under key 'h'. For example:

```
g = dgl.graph([(0, 1), (1, 2), (1, 0)], 'user', 'like', 'movie')
g.ndata['h'] = {'user': torch.zeros(2, 5), 'movie': torch.zeros(3, 5)}
```

Related issue #1629 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/heterograph.py [hunk 1]
```diff
@@ -715,7 +715,10 @@ def dstnodes(self):
     def ndata(self):
         """Return the data view of all the nodes.
 
-        **Only works if the graph has one node type.**
+        If the graph has only one node type. The g.ndata directly represents the
+        node data.
+        If the graph has multiple node type. The g.ndata is a dictionary of node
+        types and their corresponding node data.
```
Comment (mufeili) [inline python/dgl/heterograph.py:721]: How about "If the graph has multiple node types, then g.ndata[feat][node_type]" gives the features associated with key `feat` for `node_type`-typed nodes?

File: python/dgl/heterograph.py [hunk 2]
```diff
@@ -727,24 +730,41 @@ def ndata(self):
...
     @property
     def srcdata(self):
         """Return the data view of all nodes in the SRC category.
 
         Only works if the graph is either
 
-        * Uni-bipartite and has one node type in the SRC category.
+        * Uni-bipartite
+
+        * Non-uni-bipartite (in this case identical to :any:`DGLHeteroGraph.ndata`)
```
Comment (mufeili) [inline python/dgl/heterograph.py:762]: I don't think I fully get it. By non-uni-bipartite, do you mean an arbitrary graph other than a bipartite graph?

File: python/dgl/view.py [hunk 3]
```diff
@@ -345,31 +384,72 @@ def __call__(self, *args, **kwargs):
...
+        if self._multi_etype:
+            assert isinstance(val, dict), \
+                'Current HeteroEdgeDataView has multiple edge types, ' \
+                'please passing the edge type and the corresponding data through a dict.'
+
+            for (etype, data) in val.items():
+                etid = self._graph.get_etype_id(etype)
+                self._graph._set_e_repr(etid, self._edges, {key : data})
+        else:
+            self._graph._set_e_repr(self._etid, self._edges, {key : val})
```
Comment (mufeili) [inline python/dgl/view.py:420]: I think we also need to check if val is a dictionary here and raise an error if so.

File: python/dgl/heterograph.py (hunk 1 above)
Comment (classicsong) [reply python/dgl/heterograph.py:721]: > How about "If the graph has multiple node types, then g.ndata[feat][node_type] gives the features associated with key `feat` for `node_type`-typed nodes?"

This one is better.

File: python/dgl/heterograph.py (hunk 2 above)
Comment (classicsong) [reply python/dgl/heterograph.py:762]: For Uni-bipartite, src and dst are disjoint.
For Non-uni-bipartite, src and dst can have some nodes.

But I think there is no need to highlight these two kinds of graphs.

**Comment by BarclayII** [comment]
So we have the following behavior comparison:
| `ndata` behavior  | one node type | multiple node types |
|:-:|:--:|:--:|
| Previously | single tensor | error |
| This PR | single tensor | dict of ntype and tensor |

Correct?

**Comment by BarclayII** [comment]
Also, since `g.edge_subgraph` takes in edge IDs or dict of edge IDs but `train_mask` is actually a boolean edge feature tensor, `g.edge_subgraph(train_mask)` wouldn't work by definition anyway.

**Comment by BarclayII** [comment]
If you want to change the semantics of `edge_subgraph` to also support boolean masks, I would recommend also doing that in the PR.

**Comment by classicsong** [comment]
> So we have the following behavior comparison:
> 
> `ndata` behavior	one node type	multiple node types
> Previously	single tensor	error
> This PR	single tensor	dict of ntype and tensor
> Correct?

Correct

File: python/dgl/view.py [hunk 4]
```diff
@@ -282,33 +282,75 @@ def __call__(self, ntype=None):
...
+            return ret
+        else:
+            return self._graph._get_n_repr(self._ntid, self._nodes)[key]
 
     def __setitem__(self, key, val):
-        self._graph._set_n_repr(self._ntid, self._nodes, {key : val})
+        if self._multi_ntype:
+            assert isinstance(val, dict), \
+                'Current HeteroNodeDataView has multiple node types, ' \
+                'please passing the node type and the corresponding data through a dict.'
```
Comment (jermainewang) [inline python/dgl/view.py:309]: Can we allow this by setting the feature to all types.

File: python/dgl/view.py [hunk 5]
```diff
@@ -282,33 +282,75 @@ def __call__(self, ntype=None):
...
+
+            for (ntype, data) in val.items():
+                ntid = self._graph.get_ntype_id(ntype)
+                self._graph._set_n_repr(ntid, self._nodes, {key : data})
+        else:
+            assert isinstance(val, dict) is False, \
+                'The HeteroNodeDataView has only one node type. ' \
+                'please pass a tensor directly'
```
Comment (jermainewang) [inline python/dgl/view.py:317]: The check is strange. I think you want to check `val` is of tensor type instead.

File: python/dgl/heterograph.py [hunk 6]
```diff
@@ -727,24 +730,35 @@ def ndata(self):
...
-        ntype = self.ntypes[0]
-        return HeteroNodeDataView(self, ntype, ntid, ALL)
+        if len(self.ntypes) == 1:
+            ntid = self.get_ntype_id(None)
+            ntype = self.ntypes[0]
+            return HeteroNodeDataView(self, ntype, ntid, ALL)
+        else:
+            ntids = [self.get_ntype_id(ntype) for ntype in self.ntypes]
+            ntypes = self.ntypes
+            return HeteroNodeDataView(self, ntypes, ntids, ALL)
```
Comment (jermainewang) [inline python/dgl/heterograph.py:751]: I feel the codes can be simplified. Here, you just do
```python
ntids = [self.get_ntype_id(ntype) for ntype in self.ntypes]
ntypes = self.ntypes
return HeteroNodeDataView(self, ntypes, ntids, ALL)
```
no matter what. And in `HeteroNodeDataView`, you retrieve the data using the `ntids` and feature name. Before you return the data, check if the dictionary contains only one item and return the value if so.

File: python/dgl/view.py (hunk 4 above)
Comment (classicsong) [reply python/dgl/view.py:309]: Different types may have different shape. How can we know who is who in this case?

File: python/dgl/view.py (hunk 5 above)
Comment (classicsong) [reply python/dgl/view.py:317]: This assert is to give hint to someone if he pass a dict to a graph with only one edge type.

File: python/dgl/heterograph.py (hunk 6 above)
Comment (classicsong) [reply python/dgl/heterograph.py:772]: > ```python
> HeteroNodeDataView
> ```

This maybe in another PR. A lots of place use HeteroNodeDataView.

File: python/dgl/view.py (hunk 5 above)
Comment (jermainewang) [reply python/dgl/view.py:316]: So the following is not allowed?
```python
g = ...  # some homograph with node type 'user'
g.ndata['h'] = {'user': user_feat}
```

File: python/dgl/heterograph.py (hunk 6 above)
Comment (jermainewang) [reply python/dgl/heterograph.py:772]: ok

**Comment by jermainewang** [comment]
LGTM other than one small comment.

File: python/dgl/view.py (hunk 5 above)
Comment (classicsong) [reply python/dgl/view.py:316]: Yes.
For homograph, only use tensor, otherwise this may be confusion, when there is only one node type, one can use dict and also tensor.
